src/hypercall.rs

```diff
@@ -77,21 +79,90 @@ pub unsafe fn address_to_hypercall(
-/// TODO: UNSAFE AS *%@#. It has to be checked that the VM is allowed to unlink that file!
-pub fn unlink(mem: &MmapMemory, sysunlink: &mut UnlinkParams) {
-	unsafe {
-		sysunlink.ret = libc::unlink(mem.host_address(sysunlink.name).unwrap() as *const i8);
+pub fn unlink(mem: &MmapMemory, sysunlink: &mut UnlinkParams, file_map: &mut UhyveFileMap) {
+	let requested_path = mem.host_address(sysunlink.name).unwrap() as *const i8;
+	let guest_path = unsafe { CStr::from_ptr(requested_path) }.to_str();
+	if let Ok(guest_path) = guest_path {
+		// Rust deals in UTF-8. C doesn't provide such a guarantee.
+		// In that case, converting a CStr to str will return a Utf8Error.
+		//
+		// See: https://nrc.github.io/big-book-ffi/reference/strings.html
+		if let Some(host_path) = file_map.get_host_path(guest_path) {
+			// We can safely unwrap here, as host_path.as_bytes will never contain internal \0 bytes
+			// As host_path_c_string is a valid CString, this implementation is presumed to be safe.
+			let host_path_c_string = CString::new(host_path.as_bytes()).unwrap();
+
+			unsafe {
+				sysunlink.ret = libc::unlink(host_path_c_string.as_c_str().as_ptr());
+			}
+		} else {
+			error!("The kernel requested to unlink() an unknown path ({guest_path}): Rejecting...");
+			sysunlink.ret = -ENOENT;
+		}
+	} else {
+		error!("The kernel requested to open() a path that is not valid UTF-8. Rejecting...");
+		sysunlink.ret = -EIO;
 	}
 }
 
 /// Handles an open syscall by opening a file on the host.
-pub fn open(mem: &MmapMemory, sysopen: &mut OpenParams) {
-	unsafe {
-		sysopen.ret = libc::open(
-			mem.host_address(sysopen.name).unwrap() as *const i8,
-			sysopen.flags,
-			sysopen.mode,
-		);
+pub fn open(
+	mem: &MmapMemory,
+	sysopen: &mut OpenParams,
+	file_map: &mut UhyveFileMap,
+	temp_dir_base: &TempDir,
+) {
+	// TODO: Keep track of file descriptors internally, just in case the kernel doesn't close them.
+	let requested_path_ptr = mem.host_address(sysopen.name).unwrap() as *const i8;
+	let guest_path = unsafe { CStr::from_ptr(requested_path_ptr) }.to_str();
+	let mut flags = sysopen.flags & ALLOWED_OPEN_FLAGS;
+
+	if let Ok(guest_path) = guest_path {
+		// Rust deals in UTF-8. C doesn't provide such a guarantee.
+		// In that case, converting a CStr to str will return a Utf8Error.
+		//
+		// See: https://nrc.github.io/big-book-ffi/reference/strings.html
+		let host_path_option = file_map.get_host_path(guest_path);
+		if let Some(host_path) = host_path_option {
+			// We can safely unwrap here, as host_path.as_bytes will never contain internal \0 bytes
+			// As host_path_c_string is a valid CString, this implementation is presumed to be safe.
+			let host_path_c_string = CString::new(host_path.as_bytes()).unwrap();
+
+			unsafe {
+				sysopen.ret = libc::open(
+					host_path_c_string.as_c_str().as_ptr(),
+					sysopen.flags,
+					sysopen.mode,
+				);
+			}
```

Comment:
I have the feeling that rustfmt threats the following a little nice:
```suggestion
			sysopen.ret = unsafe {
				 libc::open(
					host_path_c_string.as_c_str().as_ptr(),
					sysopen.flags,
					sysopen.mode,
				)
			};
```